design/als-design.md

```diff
@@ -0,0 +1,67 @@
+# AccessLog Service (ALS) Support
+
+Status: Draft
+
+## Abstract
+
+This document proposed a design for supporting Envoy's AccessLog Service capabilities in Contour.
+
+## Background
+
+Envoy has the ability to send rich access logs to access logging sinks.
+With AccessLog Service (ALS) enabled, Envoy can send access log messages using gRPC streams to a gRPC access log service. 
+
+This document focuses on configuring access log sinks and enabling Envoy to send access log messages via Contour.
+
+## Goals
+- Enable users to specify ALS configuration
+- Enable users to send HTTP access logs for specific listeners 
+
+## Non Goals
+- TCP Access Log Support
+- As part of first version, communicate between Envoy and ALS sinks will be insecure
+- As part of first version, users can only specify a single ALS sink.
+
+## High-Level Design
+Contour will add support for Envoy's HTTP Access Log Service.
+ALS information will be specified as a [static cluster for Envoy](https://github.com/envoyproxy/envoy/blob/main/source/common/grpc/async_client_manager_impl.cc#L44).
+Contour controller will add details about ALS sink to `access_log` field for each HTTPConnectionManager filter.
+
+## Detailed Design
+To enable Envoy to send access logs to a sink; a static cluster needs to be set for Envoy that will have details about ALS sink.
+There are two ways to add this configuration:
+- ExtensionService
+- Contour Configuration file
+We have chosen to use ExtensionService to enable users to configure ALS sink details.
+
+### ExtensionService
+Contour administrator/operator can add an ExtensionService with ALS configuration details.
+As part of `contour bootstrap`; the ExtensionService will be used to configure a static ALS cluster for Envoy.
+
+An annotation `projectcontour.io/als-admitted: true` needs to be added to the particular ExtensionService.
```

Comment:
Doing ExtensionService like this does mean that there can only ever be one ALS per cluster, and we will need to build a totally different mechanism to do more than one, which seems not ideal.

On the other hand, I suspect that most people doing this are probably going to want to send to a single central log sink anyway, or to localhost. @bgagnon is currently using ALS through some clever tricks, maybe he can check out this proposal.